### PR TITLE
feat: Add health endpoint

### LINKED ISSUE #2274: Improve Health endpoint for API to actually look at the health instead of just returning 200 without looking at the functioning of the chain

### PR DESCRIPTION
**Pull request by howjmay**
close #2274

### CODE REVIEW

**Comment by howjmay** [comment]
I haven't set the correct route

File: packages/metrics/chain_messages.go [hunk 1]
```diff
@@ -58,6 +58,11 @@ type TxInclusionStateMsg struct {
 	State string
 }
 
+type ChainConfirmedState struct {
```
Comment (kape1395) [inline packages/metrics/chain_messages.go]: I think this, and related functions should be not needed.

File: packages/metrics/chain.go [hunk 2]
```diff
@@ -713,3 +716,7 @@ func (m *ChainMetricsProvider) OutPullTxInclusionState() IMessageMetric[iotago.T
 func (m *ChainMetricsProvider) OutPullOutputByID() IMessageMetric[iotago.OutputID] {
 	return m.outPullOutputByIDMetrics
 }
+
+func (m *ChainMetricsProvider) MaxChainConfirmedStateLag() uint32 { // TODO: Call it from the health endpoint.
```
Comment (jorgemmsilva) [inline packages/metrics/chain.go]:
```suggestion
func (m *ChainMetricsProvider) MaxChainConfirmedStateLag() uint32 {
```
TODO can be removed?

File: packages/webapi/api.go [hunk 3]
```diff
@@ -33,8 +33,21 @@ import (
...
+func AddHealthEndpoint(server echoswagger.ApiRoot, chainService interfaces.ChainService, metricsService interfaces.MetricsService) {
+	server.GET("/health", func(e echo.Context) error {
+		chainIDs, err := chainService.GetAllChainIDs()
+		if err != nil {
+			return e.String(http.StatusInternalServerError, fmt.Sprintf("failed to get all chain IDs: %s", err))
+		}
+
+		for _, chainID := range chainIDs {
+			lag := metricsService.GetMaxChainConfirmedStateLag(chainID)
+			if lag > 2 {
```
Comment (jorgemmsilva) [inline packages/webapi/api.go]: this `2` should probably be a configurable parameter

File: packages/webapi/api.go [hunk 4]
```diff
@@ -33,8 +33,21 @@ import (
...
-func addHealthEndpoint(server echoswagger.ApiRoot) {
-	server.GET("/health", func(c echo.Context) error { return c.NoContent(http.StatusOK) }).
+func AddHealthEndpoint(server echoswagger.ApiRoot, chainService interfaces.ChainService, metricsService interfaces.MetricsService) {
+	server.GET("/health", func(e echo.Context) error {
+		chainIDs, err := chainService.GetAllChainIDs()
+		if err != nil {
+			return e.String(http.StatusInternalServerError, fmt.Sprintf("failed to get all chain IDs: %s", err))
+		}
+
+		for _, chainID := range chainIDs {
```
Comment (jorgemmsilva) [inline packages/webapi/api.go]: could we use the `MaxLag` here?

File: packages/webapi/api.go [hunk 5]
```diff
@@ -33,8 +33,21 @@ import (
 
 const APIVersion = 1
 
-func addHealthEndpoint(server echoswagger.ApiRoot) {
-	server.GET("/health", func(c echo.Context) error { return c.NoContent(http.StatusOK) }).
+func AddHealthEndpoint(server echoswagger.ApiRoot, chainService interfaces.ChainService, metricsService interfaces.MetricsService) {
+	server.GET("/health", func(e echo.Context) error {
+		chainIDs, err := chainService.GetAllChainIDs()
+		if err != nil {
+			return e.String(http.StatusInternalServerError, fmt.Sprintf("failed to get all chain IDs: %s", err))
```
Comment (jorgemmsilva) [inline packages/webapi/api.go]: it's safe to panic in case of an internal error, it should be cached and logged by the middleware

**Review by jorgemmsilva — APPROVED** [review]

File: packages/webapi/api.go (hunk 5 above)
Comment (howjmay) [reply packages/webapi/api.go]: Um. Actually this part I have a question. We need to return an error if the node is unhealthy, but what is the error that we are going to return?

Comment (jorgemmsilva) [reply packages/webapi/api.go]: http status 500 seems ok

**Review by jorgemmsilva — APPROVED** [review]

**Pull request merged**